### PR TITLE
Razor file compilation

### LINKED ISSUE #15505: MvcRazorCompileOnBuild is worth to mention
I've build multi-modular project where each module is Razor Class Libraries (RCL) which containing bunch of views. Every time when I edit cshtml file and refresh the browser the view is recompiled. It takes too much time.

Fortunately &lt;MvcRazorCompileOnBuild&gt;false&lt;/MvcRazorCompileOnBuild&gt; switches off recompilation but changes are instantly visible in the browser. Thanks for that. I hope it newer will be removed. :)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8a83deb2-e035-bb66-be4f-68709bcad341
* Version Independent ID: 16438f35-f35d-c097-0903-ec17e83e051e
* Content: [Razor file compilation in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/view-compilation?view=aspnetcore-3.0#feedback)
* Content Source: [aspnetcore/mvc/views/view-compilation.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/views/view-compilation.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### LINKED ISSUE #16420: Under "Conditionally enable runtime compilation" Env is coming as NULL
Documentation does not say where to assign `Env`. 

Just following the documentation as as is under "Conditionally enable runtime compilation" throws a null reference exception when trying to use `Env.IsDevelopment()`.

I had to assign `Env` as below

```
public Startup(IConfiguration configuration, IWebHostEnvironment env) {
    Configuration = configuration;
    Env = env;
}
```

Then I am able to use `Env.IsDevelopment()` without throwing a null reference exception.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8a83deb2-e035-bb66-be4f-68709bcad341
* Version Independent ID: 16438f35-f35d-c097-0903-ec17e83e051e
* Content: [Razor file compilation in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/view-compilation?view=aspnetcore-3.0#feedback)
* Content Source: [aspnetcore/mvc/views/view-compilation.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/views/view-compilation.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #16420
Fixes #15505 

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/views/view-compilation?view=aspnetcore-3.1&branch=pr-en-us-17709)

Prefer bullets to numbers unless customer is likely to do the steps out of order, and order is critical. That's a rare case. When order doesn't matter, as in this case, use bullets. For multi-step instructions that do need order, bullets are better anyway. Customer will naturally follow the bullets in order. Easier to write/maintain and it looks better.

### CODE REVIEW

File: aspnetcore/mvc/views/view-compilation.md [hunk 1]
```diff
@@ -47,37 +47,23 @@ Runtime compilation can be enabled such that it's only available for local devel
...
-        if (Env.IsDevelopment())
-        {
-            builder.AddRazorRuntimeCompilation();
-        }
-    #endif
-
-        // code omitted for brevity
-    }
-    ```
+[!code-csharp[](~/mvc/views/view-compilation/sample/Startup.cs?name=snippet)]
```
Comment (scottaddie) [inline aspnetcore/mvc/views/view-compilation.md:58]: Indent this code snippet to visually associate it with the text above.

File: aspnetcore/mvc/views/view-compilation.md [hunk 2]
```diff
@@ -23,9 +23,9 @@ Build- and publish-time compilation of Razor files is enabled by default by the
 
 To enable runtime compilation for all environments and configuration modes:
 
-1. Install the [Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation/) NuGet package.
```
Comment (scottaddie) [inline aspnetcore/mvc/views/view-compilation.md:26]: Order is important here. Stick with the numbered list.

File: aspnetcore/mvc/views/view-compilation.md [hunk 3]
```diff
@@ -47,37 +47,23 @@ Runtime compilation can be enabled such that it's only available for local devel
 
 To enable runtime compilation based on the environment and configuration mode:
 
-1. Conditionally reference the [Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation/) package based on the active `Configuration` value:
```
Comment (scottaddie) [inline aspnetcore/mvc/views/view-compilation.md:50]: Same recommendation as above. These are sequential items, so a numbered list is what the style guide prescribes.

File: aspnetcore/mvc/views/view-compilation.md [hunk 4]
```diff
@@ -47,37 +47,23 @@ Runtime compilation can be enabled such that it's only available for local devel
...
-    ```
+[!code-csharp[](~/mvc/views/view-compilation/sample/Startup.cs?name=snippet)]
 
 ## Additional resources
 
+* [RazorCompileOnBuild and RazorCompileOnPublish](xref:razor-pages/sdk#properties) properties.
 * <xref:razor-pages/index>
 * <xref:mvc/views/overview>
 * <xref:razor-pages/sdk>
+* See the [runtimecompilation sample on GitHub](https://github.com/aspnet/samples/tree/master/samples/aspnetcore/mvc/runtimecompilation) for a sample that shows making `RuntimeCompilation` work across projects.
```
Comment (scottaddie) [inline aspnetcore/mvc/views/view-compilation.md:66]:
```suggestion
* See the [runtimecompilation sample on GitHub](https://github.com/aspnet/samples/tree/master/samples/aspnetcore/mvc/runtimecompilation) for a sample that shows making runtime compilation work across projects.
```

File: aspnetcore/mvc/views/view-compilation.md (hunk 2 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/views/view-compilation.md:26]: We've never had anyone complain, hey I did these out of order and it was sub-optimal in VS. That's about the only time it's even possible to be sub-optimal. Folks naturally do these in order, and if they don't, it's really not a problem. Looks much better with bullets. Reverted anyway

**Pull request merged**